deflect/qt/QuickRenderer.h

```diff
@@ -0,0 +1,147 @@
+/*    BUSINESS INTERRUPTION) HOWEVER CAUSED  AND ON ANY THEORY OF    */
+/*    LIABILITY, WHETHER  IN CONTRACT, STRICT  LIABILITY, OR TORT    */
+/*    (INCLUDING NEGLIGENCE OR OTHERWISE)  ARISING IN ANY WAY OUT    */
+/*    OF  THE  USE OF  THIS  SOFTWARE,  EVEN  IF ADVISED  OF  THE    */
+/*    POSSIBILITY OF SUCH DAMAGE.                                    */
+/*                                                                   */
+/* The views and conclusions contained in the software and           */
+/* documentation are those of the authors and should not be          */
+/* interpreted as representing official policies, either expressed   */
+/* or implied, of Ecole polytechnique federale de Lausanne.          */
+/*********************************************************************/
+
+#ifndef QUICKRENDERER_H
+#define QUICKRENDERER_H
+
+#include <QMutex>
+#include <QObject>
+#include <QOpenGLFramebufferObject>
+#include <QWaitCondition>
+
+QT_FORWARD_DECLARE_CLASS(QOffscreenSurface)
+QT_FORWARD_DECLARE_CLASS(QOpenGLContext)
+QT_FORWARD_DECLARE_CLASS(QQuickRenderControl)
+QT_FORWARD_DECLARE_CLASS(QQuickWindow)
+QT_FORWARD_DECLARE_CLASS(QSurface)
+
+namespace deflect
+{
+namespace qt
+{
+
+/**
+ * Renders the QML scene from the given window using QQuickRenderControl onto
+ * the surface of the window or to an offscreen surface. Note that this object
+ * needs to be moved to a seperate (render)thread to function properly.
+ *
+ * Inspired by http://doc.qt.io/qt-5/qtquick-rendercontrol-window-multithreaded-cpp.html
+ */
+class QuickRenderer : public QObject
+{
+    Q_OBJECT
+
+public:
+    /**
+     * After construction, move the object to a dedicated render thread before
+     * calling any other function.
+     *
+     * @param quickWindow the window to render into if not @sa offscreen
+     * @param renderControl associated with the QML scene of quickWindow to
+     *                      trigger the actual rendering
+     * @param offscreen render into an offscreen surface rather than the
+     *                  quickWindow. It creates an FBO internally to hold the
+     *                  rendered pixels.
+     */
+    QuickRenderer( QQuickWindow& quickWindow,
+                   QQuickRenderControl& renderControl,
+                   bool offscreen = false );
+
+    /** @return OpenGL context used for rendering; lives in render thread. */
+    QOpenGLContext* context() { return _context; }
+
+    /**
+     * @return nullptr if !offscreen, otherwise accessible in afterRender();
+     *         lives in render thread.
+     */
+    QOpenGLFramebufferObject* fbo() { return _fbo; }
+
+    /**
+     * To be called from GUI/main thread to trigger rendering.
+     *
+     * This call is blocking until sync() is done in render thread.
+     */
+    void render();
+
+signals:
+    /** Emitted at the end of render(). Originates from render thread. */
+    void afterRender();
```

Comment:
From the changes in Tide, I see that it is necessary to connect this signal to do the swapBuffer when not in offscreen mode, right? Maybe add that to the doc?